autopilot/contractor.go

```diff
@@ -0,0 +1,536 @@
+package autopilot
+
+import (
+	"errors"
+	"math"
+	"sync"
+	"time"
+
+	"go.sia.tech/renterd/bus"
+	"go.sia.tech/renterd/internal/consensus"
+	rhpv2 "go.sia.tech/renterd/rhp/v2"
+	"go.sia.tech/renterd/worker"
+	"go.sia.tech/siad/types"
+)
+
+const (
+	// estimatedFileContractTransactionSetSize is the estimated blockchain size
+	// of a transaction set between a renter and a host that contains a file
+	// contract.
+	//
+	// TODO: move this constant to a separate package along with all others
+	estimatedFileContractTransactionSetSize = 2048
+
+	// stopTimeout is the amount of time we wait on stop before erroring out
+	// indicating an unclean shutdown
+	stopTimeout = time.Minute
+)
+
+type contractor struct {
+	ap                   *Autopilot
+	contractLoopInterval time.Duration
+
+	wg       sync.WaitGroup
+	stopChan chan struct{}
+	wakeChan chan struct{}
+}
+
+func newContractor(ap *Autopilot, contractLoopInterval time.Duration) *contractor {
+	return &contractor{
+		ap:                   ap,
+		contractLoopInterval: contractLoopInterval,
+		stopChan:             make(chan struct{}),
+		wakeChan:             make(chan struct{}),
+	}
+}
+
+func (c *contractor) run() {
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		c.contractLoop()
+	}()
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		c.triggerLoop()
+	}()
+}
+
+func (c *contractor) stop() error {
+	close(c.stopChan)
+
+	doneChan := make(chan struct{})
+	go func() {
+		c.wg.Wait()
+		close(doneChan)
+	}()
+	select {
+	case <-doneChan:
+	case <-time.After(stopTimeout):
+		return errors.New("unclean contractor shutdown")
+	}
+	return nil
+}
+
+func (c *contractor) triggerLoop() {
+	for {
+		select {
+		case <-c.stopChan:
+			return
+		case <-time.After(c.contractLoopInterval):
+			c.wake()
```

Comment:
this is definitely a bit weird. I can see why it happened -- you're trying to factor out the `time`-based triggers like I recommended. I'll have to think about how to avoid this.